src/app/user/register/register.component.ts

```diff
@@ -9,38 +11,126 @@ import { Router } from '@angular/router';
-    return true;
+  passwordsValid(control: AbstractControl): { [key: string]: boolean } | null {
+    this.errorMessage = '';
+    const password1 = control.get('userPasswordControl')?.value.toString();
+    const password2 = control.get('userPasswordConfirmControl')?.value.toString();
+    
+    if(password1.indexOf(' ') >= 0) {
+      this.errorMessage = ErrorMessages.passwordsHasSpaces;
+      // this.errorMessage = 'Passowrd may not have spaces';
+      return {'passwordsInvalid:': true};
+    }
+
+    // Uncomment this when we're ready to test for 8 or more charachters,
+    // with at least one letter, number and symbol.
+    // else if(!/^(?=.*[A-Za-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$/.test(password1)){
+    //   this.errorMessage = ErrorMessages.passwordDoesNotMeetRegEx;
+    //   return { 'passwordsInvalid': true };
+    // }
+
+    // console.log(control.get('userPasswordControl'))
+    else if ((password1 !== password2)) {
+      this.errorMessage = ErrorMessages.passwordsMustMatch;
+      return { 'passwordsInvalid': true };
+    }
+    return null;
   }
 
   register(userEmail: string, userPassword: string, userConfirmPassword: string): void {
-    if (userPassword === userConfirmPassword && this.userPasswordIsValid()) {
-      // RAFA: Add validation here (if passswords don't match, throw error)
+    this.registerErrorMessage = '';
+    // Double checking that passwords are equal to each other but redundant as its
+    // alread checked by the form --> (this.userPasswordsGroup.valid)
+    const passwordsEqual = userPassword === userConfirmPassword;
+
+    if (this.userPasswordsGroup.valid && this.userEmailControl.valid && passwordsEqual) {
+      console.log("Creating user!");
       this.auth.createUserWithEmailAndPassword(userEmail, userPassword)
       .then(() => {
-        console.log(userEmail);
-        console.log(userPassword);
-        console.log(userConfirmPassword);
+        console.log('Created user: ', userEmail);
         this.router.navigateByUrl('/');
       })
       .catch((error) => {
+        this.registerErrorMessage = error.message;
         // Use error in validation
-        console.error(error);
+        //console.error(error);
       });
     }
-
-    this.registerError = true;
+    else {
+      this.registerErrorMessage = ErrorMessages.makeSureInputFieldsValid;
+      this.registerError = true;
+    }
   }
+}
 
-
+/**
+ * Custom ErrorStateMatcher which returns true (error exists) when the parent form group is invalid and the control has been touched
+ */
+export class ConfirmPasswordsMatch implements ErrorStateMatcher {
+  isErrorState(control: FormControl | null, form: FormGroupDirective | NgForm | null): boolean {
+      const parentValid = control!.parent!.valid;
+      const firstPasswordTyped = control!.parent!.get('userPasswordControl')!.dirty;
+      
+      return !parentValid && firstPasswordTyped      
+  }
 }
 
+export const ErrorMessages: { [key: string]: string } = {
```

Comment:
what does the syntax mean here? dont need to fix, just curious. is that just the constructor shorthand

Reply:
Its creating a dictionary of where the key is a string and the value is also a string. The dictionary would have the following structure:

`
{
someStringKey: "key value"
}
`